sources/packages/backend/apps/api/src/services/education-program/education-program.service.ts

```diff
@@ -803,6 +819,155 @@ export class EducationProgramService extends RecordDataModelService<EducationPro
+            }),
+        "totalOfferings",
+      )
+      .innerJoin("programs.institution", "institution")
+      .innerJoin(
+        InstitutionLocation,
+        "location",
+        "institution.id = location.institution.id",
+      )
+      .where("programs.institution.id = :institutionId", { institutionId });
+
+    // This queryParams is for getRawCount, which is different from the
+    // query parameter assigned to paginatedProgramQuery like
+    // paginationOptions.searchCriteria or institutionId
+    // queryParams should follow the order/index.
+    const queryParams: any[] = [...offeringTypes, institutionId];
+
+    return { programQuery, queryParams };
+  }
+
+  /**
+   * Prepares the paginated query and the count.
+   * @param paginatedProgramQuery paginated program query.
+   * @param paginationOptions pagination options.
+   * @param queryParams query parameters.
+   * @returns paginated result along with the count.
+   */
+  private async preparePaginatedProgramQuery(
+    paginatedProgramQuery: SelectQueryBuilder<EducationProgram>,
+    paginationOptions: PaginationOptions,
+    queryParams: any[],
+  ): Promise<PaginatedResults<EducationProgramsSummary>> {
+    // For getting total raw count before pagination.
+    const sqlQuery = paginatedProgramQuery.getSql();
+
+    if (paginationOptions.pageLimit) {
+      paginatedProgramQuery.limit(paginationOptions.pageLimit);
+    }
+
+    paginatedProgramQuery.offset(
+      paginationOptions.page * paginationOptions.pageLimit,
+    );
+
+    // sort
+    if (
+      paginationOptions.sortField === "programStatus" &&
+      paginationOptions.sortOrder === "DESC"
+    ) {
+      paginatedProgramQuery.orderBy(
+        `CASE          
+          WHEN programs.programStatus = '${ProgramStatus.Pending}' and programs.isActive = true THEN ${SortPriority.Priority1}
+          WHEN programs.isActive = false THEN ${SortPriority.Priority2}
+          WHEN programs.programStatus = '${ProgramStatus.Declined}' and programs.isActive = true THEN ${SortPriority.Priority3}
+          WHEN programs.programStatus = '${ProgramStatus.Approved}' and programs.isActive = true THEN ${SortPriority.Priority4}
+          ELSE ${SortPriority.Priority5}
+        END`,
+      );
+    } else if (
+      paginationOptions.sortField === "programStatus" &&
+      paginationOptions.sortOrder === "ASC"
+    ) {
+      paginatedProgramQuery.orderBy(
+        `CASE          
+          WHEN programs.programStatus = '${ProgramStatus.Approved}' and programs.isActive = true THEN ${SortPriority.Priority1}
+          WHEN programs.programStatus = '${ProgramStatus.Declined}' and programs.isActive = true THEN ${SortPriority.Priority2}
+          WHEN programs.isActive = false THEN ${SortPriority.Priority3}
+          WHEN programs.programStatus = '${ProgramStatus.Pending}' and programs.isActive = true THEN ${SortPriority.Priority4}
+          ELSE ${SortPriority.Priority5}
+        END`,
+      );
```

Comment:
Since the requirement is to have this ordered alphabetically, can the first part be simplified as below?
```ts
// Sort
if (paginationOptions.sortField === "programStatus") {
  paginatedProgramQuery.orderBy(
    `CASE WHEN programs.isActive = false THEN '${INACTIVE_PROGRAM_STATUS}' ELSE programs.programStatus :: text end`,
    paginationOptions.sortOrder,
  );
} else if (paginationOptions.sortField && paginationOptions.sortOrder) {
  paginatedProgramQuery.orderBy(
    sortProgramsColumnMap(paginationOptions.sortField),
    paginationOptions.sortOrder,
  );
}
```